syntax = "proto3";

package xla.poplarplugin;

message IpuOptions {

  // Structure for setting an idividual Poplar OptionFlag for OptionFlags
  message PoplarOptionFlag {
    string option = 1;
    string value = 2;
  };

  // Options controlling the software IPU model (see IPUModel in poplar)
  message IpuModelConfig {
    bool enable_ipu_model = 1;
    bool compile_ipu_code = 2;
  };
  IpuModelConfig ipu_model_config = 1;

  // Options to control IPU profiling
  message Profiling {
    // Set to enable compilation trace
    bool enable_compilation_trace = 1;

    // Set to have the poplar reports in text, not JSON
    bool enable_poplar_reports_text = 2;

    // Set to enable I/O trace
    bool enable_io_trace = 3;

    // Set to enable execution trace
    bool enable_execution_trace = 4;

    // Report every N'th execution (0=once only)
    int64 report_every_nth_execution = 5;

    // Options for controlling profile report generation
    repeated PoplarOptionFlag options = 6;

    // Enable IPU trace events - indpendently of poplar reporting
    bool enable_ipu_trace_events = 7;

    // Set to have the poplar reports in CBOR, not JSON
    bool enable_poplar_reports_cbor = 8;

    // When non-empty, reports will be written into this directory, instead of
    // into the Tensorboard events
    string report_directory = 9;

    // Reports over this size (in bytes) will be discarded
    int64 max_report_size = 10;
  };
  Profiling profiling = 2;

  // Options to control Poplar compilation
  repeated PoplarOptionFlag compilation_options = 3;

  // Options controlling the configuration of each IPU device
  message DeviceConfig {
    // Set one of these fields to determine the way to select IPUs
    oneof selection {
      // Set this to the number of IPUs which are required.  The system will
      // acquire a free IPU configuration containing this many physical IPUs.
      int32 auto_count = 1;

      // Set this to acquire a specific IPU hardware configuration.
      int32 cfg_index = 2;
    }

    // Number of replicas (should be a divisor of the number of IPUs
    int32 num_replicas = 3;
  };
  repeated DeviceConfig device_config = 4;

  // Options to control the memory size and speed of the execution trade-off.
  message SpeedSizeConfig {
    // Option controlling whether to rearrange streamed copies on the host
    bool always_rearrange_copies_on_the_host = 1;

    // Option to disable caching of sub-graphs containing convolutions
    bool disable_graph_convolution_caching = 2;

    // Option to enable re-computation of non linearities.
    bool recompute_norm_inputs = 3;
    bool has_recompute_norm_inputs = 4;

    // Merge all streaming reads into a block to reduce the host-sync round-trip
    // count.  This may decrease the device->host read latency at the expense of
    // having more live tensors in memory on the device.
    bool merge_infeed_io_copies = 5;
  }
  SpeedSizeConfig speed_size_config = 5;

  // Options controlling the configuration of convolutions
  repeated PoplarOptionFlag convolution_options = 6;

  // Options controlling the configuration of pooling operations
  repeated PoplarOptionFlag pooling_options = 7;

  // Retain control dependencies from the Tensorflow graph
  bool retain_control_dependencies = 8;

  // The maximum number of bytes to wait before scheduling an all-reduce
  int64 max_cross_replica_sum_buffer_size = 9;

  message FloatingPointBehaviour {
    // When set, the flags below will be used to configure the hardware at the
    // start of program execution.
    bool flags_set = 1;

    // These flags match the ones provided by the poplar::FloatingPointBehaviour
    // structure
    bool inv = 2;
    bool div0 = 3;
    bool oflo = 4;
    bool esr = 5;
    bool nanoo = 6;
  }
  FloatingPointBehaviour floating_point_behaviour = 10;
};
